frame/collective/src/lib.rs

```diff
@@ -968,6 +976,48 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		});
 		num_proposals as u32
 	}
+
+	/// Ensure the correctness of the state of this pallet.
+	///
+	/// The following expectation must always apply.
+	///
+	/// ## Expectations:
+	///
+	/// Looking at proposals:
+	///
+	/// * Each hash of a proposal that is stored inside `Proposals` must have a
+	/// call mapped to it inside the `ProposalsOf` storage map.
+	/// * `ProposalCount` must always be more or equal to the number of
+	/// proposals inside the `Proposals` storage value. The reason why
+	/// `ProposalCount` can be more is because when a proposal is removed the
+	/// count is not deducted.
+	///
+	/// Looking at votes:
+	/// * The sum of aye and nay votes for a proposal can never exceed
+	///  `MaxMembers`.
+	///
+	/// Looking at members:
+	/// * The members count must never exceed `MaxMembers`.
+	#[cfg(any(feature = "try-runtime", test))]
+	fn do_try_state() -> Result<(), &'static str> {
+		Self::proposals().into_iter().for_each(|proposal| {
+			assert!(Self::proposal_of(proposal).is_some());
+		});
+
+		assert!(Self::proposals().into_iter().count() <= Self::proposal_count() as usize);
```

Comment:
Rather use `ensure!` here since this function should return an `Err`, not panic.

Reply:
But aren't we panicking in other `try_state` hooks? Looking at the nomination pools: https://github.com/paritytech/substrate/blob/8bd57418b79d2450079755c0d631168b8fad2e97/frame/nomination-pools/src/lib.rs#L3059-L3072

Reply:
Huh, weird. Then why does it return an error @kianenigma :rofl: ?

Reply:
I was never happy with the `&'static str` so I may have ignored it in a few places 🙈  not very nice of me. Please return an error and see: https://github.com/paritytech/substrate/issues/13736